### PR TITLE
Chemistry functionality remake

### PR DESCRIPTION
**Pull request by TorkelE**
Changes:
- Add docstrings to chemistry-related functions.
- Adds documentation for chemistry-related functionality.
- Simplifies internal code for chemistry-related functionality.
- Redo `@compound` syntax:
```julia
@variables t
@species C(t) O(t)
@compound CO2(t) ~ C + 2O 
```
- Enables `@compound` to see variables, e.g.
```julia
@variables t
@species C(t) O(t)
o = O
@compound CO2 ~ C + 2o
```
- Add the `compounds` macro, enabling the creation of multiple compounds simultaneously:
```julia
@variables t
@species C(t) O(t)
@compounds begin
    O2 ~ 2O
    CO2 ~ C + 2O
end
```
- Adds the `compounds` option to the DSL, enabling the declaration of compounds within it (the components must be designated as a species, and are not ifered).

### CODE REVIEW

**Comment by TorkelE** [comment]
Ok, this shoudl be ready now. I will have a look tomorrow with fresh eyes though to be sure there are no errors.

One feature that it might make sense to add is a `balance_reaction_network` function, which applied `balance_reaction` on each reaction of a `ReactionSystem`, and returns a new `ReactionSystem` with balanced reactions.

Finally, I don't think setting metadata is supported for compounds, which probably should be the case, e.g.
```julia
@compound H2_2(t) [metadata=value] = 2H
@compounds begin
    O2  [metadata=value]= 2O 
end
```

**Comment by TorkelE** [comment]
Ok, this should be ready now.

**Comment by isaacsas** [comment]
Yes, we should support metadata for compounds. 

Should we instead infer the arguments of the compound from the species that make it up?

**Comment by isaacsas** [comment]
I.e. if the species depend on time `t` the compound does, but if they depend on `s` then it does?

**Comment by TorkelE** [comment]
That's a good point. Having a compound with a different independent variable than its compositions would be weird, right? If this is indeed the case I will change it so that the compound takes its components iv (and throw an error if they have different ones.

Supporting metadata should be easy, e.g.
```julia
@compounds begin
    O2  [metadata=value] = 2O 
end
```
But if we want to also give a default value, how would we phrase that?
```julia
@compounds begin
    O2  [metadata=value] = 2O =1.0
end
```
?

**Comment by isaacsas** [comment]
Or the compound could be the union of the `ivs` of the components I guess (for mixed spatial problems).

**Comment by isaacsas** [comment]
Yeah, we should allow default values. Maybe we should use something else to indicate the formula in terms of species? 

```julia
@compounds begin
    O2  [metadata=value] = 1.0 ~ 2O
end
```
But I’m open to other suggestions.

**Comment by TorkelE** [comment]
I will have a look tomorrow at what is actually possible notations using metaprogramming. Leaning towards keeping the current notation, and then add something if default values are required.

**Comment by isaacsas** [comment]
The one reason for `~` is then it looks like an MTK equation.

**Comment by TorkelE** [comment]
With default values being stored in the metadata, would it be possible to define it there? Doesn't seem to work perfectly though, e.g. 
```julia
@variables t X(t) [defaultvalue=1.0]
```
and other alternatives I've tried fail. Stuff like this:
```
@compound CO2 = 0.0 = C + 2O 
```
would be possible.

I think that, for now, I will leave setting default values unsupported (unless we can make it work along the line like `[defaultvalue=1.0]`, in which case there is a workaround. The feature seems incredibly niche, and given that we cannot know whether anyone will actually use compounds a year from now (not to talk about setting default values for them), I don't want to sink too much time into it. If I see an opportunity to do it easily while fixing the metadata stuff I might have a go at it though.

**Comment by isaacsas** [comment]
OK, if it isn't easy to do maybe we should hold off on making an update then. I'd prefer not to potentially break the interface twice (i.e. once with this PR, and then potentially again when we fix it to support metadata and initial values).

**Comment by TorkelE** [comment]
Updated how `@compound` (and similar) have its syntax:
```julia
@compound H2O_1 ~ 2H + O
@compound H2O_2, [output=true] ~ 2H + O
@compound (H2O_3 = 1.5) ~ 2H + O
@compound (H2O_4 = 4, [output=true]) ~ 2H + O
@compound (H2O_5 = p1, [output=true]) ~ 2H + p2*O
```

Currently, I have not been able to figure out how to extract the independent variable from the rhs, so just set it to `t`. Once I've figured that out this should be ready.

**Comment by TorkelE** [comment]
Sorry, I've been juggling 10 branches back and forth and I finally messed up. I accidentally rebased this on https://github.com/SciML/Catalyst.jl/pull/713 (I meant to do that for the structural identifiability PR, which needs to be on that PR to be able to handle mm and hill functions). Talked to Avik and there doesn't seem to be a good solution more than waiting with merging this until https://github.com/SciML/Catalyst.jl/pull/713 is merged.

**Comment by TorkelE** [comment]
Updates:
- Improved internal formatting further.
- Change syntax to `@compound (H2O_2, [output=true]) ~ 2H + O` (parenthesis added).
- Permits adding ivs in the compound.
- If the components have more than one iv combined, ivs must be provided in the compound.

File: src/chemistry_functionality.jl
Comment (isaacsas) [inline src/chemistry_functionality.jl:21]:
```suggestion
@compound CO2(t) ~ C + 2O
```

File: src/chemistry_functionality.jl
Comment (isaacsas) [inline src/chemistry_functionality.jl:34]: Is this still true or do you infer it as the union? In the tutorial you say you shouldn't give arguments to the compound I think.

File: src/chemistry_functionality.jl
Comment (isaacsas) [inline src/chemistry_functionality.jl:110]: `~`

File: src/chemistry_functionality.jl
Comment (isaacsas) [inline src/chemistry_functionality.jl:290]: `~`

File: src/chemistry_functionality.jl
Comment (isaacsas) [inline src/chemistry_functionality.jl:302]: `~`

File: src/chemistry_functionality.jl
Comment (isaacsas) [inline src/chemistry_functionality.jl:314]: `~`

**Review by isaacsas — CHANGES_REQUESTED** [review]
Does balancing with reactions involving compounds of compounds work / is it tested? 

@TorkelE I'm fine with merging once you've addressed these comments -- I don't need to review again. If compounds of compounds don't work in balancing that should be mentioned in the tutorial and docstring though (or made to work).

**Comment by TorkelE** [comment]
Compound of compounds does not work. Thought it was mentioned, but will add that in (and in docstring as well).

**Comment by isaacsas** [comment]
Can we add an error message in `balance_reaction` if we detect one of the species is a compound of compounds? That should be straightforward I'd hope.

**Comment by isaacsas** [comment]
(Unless I just missed that we already give an error there.)

**Comment by TorkelE** [comment]
Checked, no there is not such error. I will add that.

**Comment by isaacsas** [comment]
OK, thanks! Once you've made your updates feel free to merge if tests pass. (They might not until we get the binomial issue in Symbolics fixed.)

File: docs/src/catalyst_functionality/chemistry_related_functionality.md
Comment (TorkelE) [reply docs/src/catalyst_functionality/chemistry_related_functionality.md:69]: Yes they can, will rephrase to make this clear.

File: docs/src/catalyst_functionality/chemistry_related_functionality.md
Comment (TorkelE) [reply docs/src/catalyst_functionality/chemistry_related_functionality.md:126]: agreed. I will raise an issue (otherwise I'm bound to forget it).

File: src/chemistry_functionality.jl
Comment (TorkelE) [reply src/chemistry_functionality.jl:290]: fixed

File: src/chemistry_functionality.jl
Comment (TorkelE) [reply src/chemistry_functionality.jl:302]: fixed

File: src/chemistry_functionality.jl
Comment (TorkelE) [reply src/chemistry_functionality.jl:314]: fixed

File: src/chemistry_functionality.jl
Comment (TorkelE) [reply src/chemistry_functionality.jl:110]: fixed

File: src/chemistry_functionality.jl
Comment (TorkelE) [reply src/chemistry_functionality.jl:34]: Arguments are required, we couldn't do the union because we wouldn't be able to infer the correct order of the ivs. Will double read through the tutorial to ensure this is written there.

File: docs/src/catalyst_functionality/chemistry_related_functionality.md
Comment (TorkelE) [reply docs/src/catalyst_functionality/chemistry_related_functionality.md:5]: yes

File: src/chemistry_functionality.jl
Comment (TorkelE) [reply src/chemistry_functionality.jl:34]: added an extra bit about this.

**Comment by TorkelE** [comment]
This is done now. Will give it a day or two to see what happens with the binomial issue before merging though (in case there is some problem with that).

**Comment by isaacsas** [comment]
We need to wait till we can run tests to completion before merging anything.

**Comment by TorkelE** [comment]
yes

**Pull request closed without merging**
